README.md
This project has moved to Sourcehut: <https://git.sr.ht/~ljc/putqr>

---

# putqr

[![Gem Version](https://badge.fury.io/rb/putqr.svg)](https://rubygems.org/gems/putqr)

Display a QR code in your terminal.

If you're using [iTerm2](https://iterm2.com) on macOS,
the QR code will be displayed as an image.

![Example: putqr https://example.com](images/example.png)

## Install

Install `putqr` from [RubyGems](https://rubygems.org/):

```
gem install putqr
```

## Usage

Run the `putqr` command with the string you want to encode, like:

```
putqr https://example.com/
```

Run the command without arguments to read the data from standard input:

```
putqr < example.txt
```

## Development

To run the latest source code, check out the Git repository:

```
git clone https://git.sr.ht/~ljc/putqr
```

Install the dependencies using Bundler:

```
gem install bundler
bundle install
```

Run the tests and the `putqr` command using Bundler:

```
bundle exec rspec
bundle exec putqr
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


<!-- ## [Unreleased] -->


## [0.2.0] - 2018-11-04

### Added
- `--[no-]image` argument (overrides the default behavior of detecting iTerm2).
- `--version` argument.

### Changed
- iTerm2 images now appear in tmux.

### Fixed
- Strip the final newline from stdin.


## [0.1.0] - 2018-11-03

### Added
- Initial release. Generate a QR code and print it in the terminal.
- Detect iTerm2 and display the QR code as an inline image.


[Unreleased]: https://git.sr.ht/~ljc/putqr/log
[0.2.0]: https://git.sr.ht/~ljc/putqr/log/v0.2.0
[0.1.0]: https://git.sr.ht/~ljc/putqr/log/v0.1.0


spec/qrcode_spec.rb
require_relative 'spec_helper'
require 'putqr'

RSpec.describe PutQR::QRCode do
  let(:qrcode) { described_class.new(content) }

  context 'with a short string of digits' do
    let(:content) { '12345' }

    it 'is valid' do
      expect(qrcode.valid?).to be true
    end
  end

  context 'with a URL' do
    let(:content) { 'https://example.com' }

    it 'is valid' do
      expect(qrcode.valid?).to be true
    end
  end

  context 'with more than the maximum input length' do
    let(:content) { 'Long string...' * 1_000 }

    it 'is invalid' do
      expect(qrcode.valid?).to be false
    end
  end
end


lib/qrcat.rb
# frozen_string_literal: true

require 'base64'
require 'rqrcode'

module QRCat
  # QRCode renderer.
  class QRCode
    attr_reader :qrcode

    def initialize(content)
      @qrcode = QRCode.generate_qrcode(content)
    end

    def valid?
      !qrcode.nil?
    end

    def render
      if ENV['TERM_PROGRAM'].start_with? 'iTerm'
        render_image_iterm2
      else
        render_ansi
      end
    end

    def render_ansi
      qrcode.as_ansi if valid?
    end

    def render_image_iterm2
      return nil unless valid?

      png = qrcode.as_png(size: 600)
      png_base64 = Base64.encode64(png.to_s).chomp
      options = 'inline=1'
      "\033]1337;File=#{options}:#{png_base64}\007"
    end

    class << self
      # Try each size until one fits
      def generate_qrcode(content)
        (min_qr_version(content.size)..40).each do |version|
          begin
            return RQRCode::QRCode.new(content, level: :h, size: version)
          rescue RQRCode::QRCodeRunTimeError
            next
          end
        end
        nil
      end

      # Skip the lower QR code versions where the input is known to be too
      # long. These figures are based on the maximum number of characters
      # that can be encoded for a numeric string with high error correction.
      def min_qr_version(input_size)
        if    input_size >= 2_524 then 36
        elsif input_size >= 1_897 then 31
        elsif input_size >=   969 then 21
        elsif input_size >=   331 then 11
        else 1
        end
      end
    end
  end
end


lib/putqr.rb
# frozen_string_literal: true

require 'base64'
require 'rqrcode'

require 'putqr/version'

module PutQR
  # Display a QR code in your terminal.
  class QRCode
    attr_reader :qrcode

    # Initialize the QR code with a string.
    def initialize(content)
      @qrcode = QRCode.generate_qrcode(content)
    end

    # Can the string be encoded as a QR code?
    def valid?
      !qrcode.nil?
    end

    # Render the QR code using the best method available for the terminal.
    # Returns a string.
    def render
      if ENV['TERM_PROGRAM'].start_with? 'iTerm'
        render_image_iterm2
      else
        render_ansi
      end
    end

    # Render the QR code using ANSI escape codes.
    # Returns a string.
    def render_ansi
      qrcode.as_ansi if valid?
    end

    # Render the QR code as an inline image.
    # Returns a string.
    def render_image
      render_image_iterm2
    end

    # Render the QR code as an inline image for iTerm2.
    # Returns a string.
    def render_image_iterm2
      return nil unless valid?

      # References:
      # https://iterm2.com/documentation-images.html
      # https://iterm2.com/utilities/imgcat

      # tmux requires some extra work for unrecognised escape code sequences
      screen = ENV['TERM'].start_with? 'screen'
      prefix = screen ? "\ePtmux;\e\e]" : "\e]"
      suffix = screen ? "\a\e\\" : "\a"

      png = qrcode.as_png(size: 600)
      png_base64 = Base64.encode64(png.to_s).chomp
      options = 'inline=1'
      "#{prefix}1337;File=#{options}:#{png_base64}#{suffix}"
    end

    def self.generate_qrcode(content)
      # Try each size until one fits
      (min_qr_version(content.size)..40).each do |version|
        begin
          return RQRCode::QRCode.new(content, level: :h, size: version)
        rescue RQRCode::QRCodeRunTimeError
          next
        end
      end
      nil
    end

    def self.min_qr_version(input_size)
      # Skip the lower QR code versions where the input is known to be too
      # long. These figures are based on the maximum number of characters
      # that can be encoded for a numeric string with high error correction.
      if    input_size >= 2_524 then 36
      elsif input_size >= 1_897 then 31
      elsif input_size >=   969 then 21
      elsif input_size >=   331 then 11
      else 1
      end
    end
  end
end
